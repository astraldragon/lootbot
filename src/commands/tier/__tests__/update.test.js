jest.mock("discord-akairo");
jest.mock("../../../models", () => {
  return {
    Tier: { update: jest.fn("findOrCreate") }
  };
});

const { Command } = require("discord-akairo");
const { Tier } = require("../../../models");
const { checkManagePermissions } = require("../../../support");
const UpdateCommand = require("../update");

describe("commands/tier/add", () => {
  beforeEach(() => {
    this.command = new UpdateCommand();

    this.args = {
      existingName: "Tier 1",
      name: null,
      color: null,
      image: null,
      weight: null,
      luckyWeight: null
    };

    this.msg = {
      guild: { id: "1234" },
      channel: {
        send: jest.fn("send").mockReturnValue(Promise.resolve())
      }
    };
  });

  it("should configure properly", () => {
    expect(Command).toMatchSnapshot();
  });

  it("should have the proper permission check", () => {
    expect(Command.mock.calls[0][1].options.permissions).toBe(
      checkManagePermissions
    );
  });

  it("should update only non-null values", async () => {
    Tier.update.mockReturnValue([1]);
    this.args = { ...this.args, name: "New Name" };
    await this.command.exec(this.msg, this.args);

    expect(Tier.update).toHaveBeenCalledWith(
      { name: this.args.name },
      { where: { name: this.args.existingName, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.name} updated`
    );
  });

  it("should update the image if provided", async () => {
    Tier.update.mockReturnValue([1]);
    this.args = { ...this.args, image: { href: "test.png" } };
    await this.command.exec(this.msg, this.args);

    expect(Tier.update).toHaveBeenCalledWith(
      { image: this.args.image.href },
      { where: { name: this.args.existingName, guild: this.msg.guild.id } }
    );

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.existingName} updated`
    );
  });

  it("should notify when the tier is not found", async () => {
    Tier.update.mockReturnValue([0]);
    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `${this.args.existingName} not found`
    );
  });

  it("should notify if any errors occur", async () => {
    Tier.update.mockImplementation(() => {
      throw new Error("test");
    });

    await this.command.exec(this.msg, this.args);

    expect(this.msg.channel.send).toHaveBeenCalledWith(
      `An error occurred updating ${this.args.existingName}`
    );
  });
});
