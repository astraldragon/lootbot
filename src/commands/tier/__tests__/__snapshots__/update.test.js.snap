// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`commands/tier/add should configure properly 1`] = `
[MockFunction] {
  "calls": Array [
    Array [
      "tier-update",
      Object {
        "aliases": Array [
          "tier-update",
          "tu",
        ],
        "args": Array [
          Object {
            "id": "existingName",
            "prompt": Object {
              "start": "What is the current name of the tier?",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "name",
            "match": "prefix",
            "prefix": "name=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new name of the tier?",
            },
            "type": "string",
          },
          Object {
            "default": null,
            "id": "color",
            "match": "prefix",
            "prefix": "color=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new color of the tier?",
            },
            "type": "color",
          },
          Object {
            "default": null,
            "id": "image",
            "match": "prefix",
            "prefix": "image=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new image of the tier?",
            },
            "type": "url",
          },
          Object {
            "default": null,
            "id": "weight",
            "match": "prefix",
            "prefix": "weight=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new weight of the tier?",
            },
            "type": "number",
          },
          Object {
            "default": null,
            "id": "luckyWeight",
            "match": "prefix",
            "prefix": "luckyWeight=",
            "prompt": Object {
              "optional": true,
              "start": "What is the new lucky weight of the tier?",
            },
            "type": "number",
          },
        ],
        "category": "Tier",
        "channelRestriction": "guild",
        "description": Object {
          "content": "Update that tier",
          "examples": Array [
            "tier-update Common name=\\"New Common\\" color=#555555 image=https://some.pathtoimage.png weight=25 luckyWeight=75",
            "tier-update \\"Super Common\\" weight=10",
          ],
          "usage": "<existingName> name=<name> color=<color> image=<image> weight=<weight> luckyWeight=<luckyWeight>",
        },
        "options": Object {
          "permissions": [Function],
        },
        "split": "sticky",
      },
    ],
  ],
}
`;
